server/render/index.js

```diff
@@ -124,13 +124,21 @@ export function render( element, key = JSON.stringify( element ), req ) {
 
 export function serverRender( req, res ) {
 	const context = req.context;
+	const { cacheQueryKeys = [] } = context;
+
 	let title,
 		metas = [],
 		links = [],
 		cacheKey = false;
 
 	if ( isSectionIsomorphic( context.store.getState() ) && ! context.user ) {
-		cacheKey = getCacheKey( context );
+		if ( isEmpty( context.query ) ) {
+			cacheKey = context.pathname;
+
+			// If we have query args, make sure we only cache if they're *all* whitelisted
+		} else if ( includes( cacheQueryKeys, Object.keys( context.query ) ) ) {
```

Comment:
This does't check whether an array contains elements from another array:

```js
includes( ['a','b','c'], Object.keys({a:'',b:'',c:''}))
// false
includes( ['a','b','c'], ['a','b','c'])
// false

// Maybe this?
every( Object.keys( context.query ), key => includes( cacheQueryKeys, key ) )
```

Maybe we can add more tests here 🙂 

See the check in #22593 https://github.com/Automattic/wp-calypso/blob/425c0b9b5207ffb9b0650a057a9203d5b5437dee/server/isomorphic-routing/index.js#L108-L112

That implementation is extremely strict in that it treats cacheQueryKeys as atomic sets of queries. That seemed necessary because providing `client_id` without `redirect_to` was an error. 
https://github.com/Automattic/wp-calypso/blob/032cfc6587430685e8f3d3677076cc832508c4c6/client/login/controller.js#L49-L54

Without wanting to over-engineer this, I think the approach here should be fine. It may be good to document that I was considering something like the following:

```js
cacheQueryKeys = [
  { required: …requiredKeys, optional: …optionalKeys },
]
```

The above really starts to look like a JSON schema… 🤔 In a future where these advanced things become necessary –did you already mention whitelisting certain values?– a schema could handle all of that in an expressive way.